Userland/Utilities/zip.cpp

```diff
@@ -54,52 +51,28 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
     Archive::ZipOutputStream zip_stream(move(file_stream));
 
     auto add_file = [&](DeprecatedString path) -> ErrorOr<void> {
-        auto canonicalized_path = LexicalPath::canonicalized_path(path);
-        auto file = TRY(Core::File::open(path, Core::File::OpenMode::Read));
-        auto file_buffer = TRY(file->read_until_eof());
-        Archive::ZipMember member {};
-        member.name = TRY(String::from_deprecated_string(canonicalized_path));
+        auto canonicalized_path = TRY(String::from_deprecated_string(LexicalPath::canonicalized_path(path)));
 
+        auto file = TRY(Core::File::open(path, Core::File::OpenMode::Read));
         auto stat = TRY(Core::System::fstat(file->fd()));
         auto date = Core::DateTime::from_timestamp(stat.st_mtim.tv_sec);
-        member.modification_date = to_packed_dos_date(date.year(), date.month(), date.day());
-        member.modification_time = to_packed_dos_time(date.hour(), date.minute(), date.second());
-
-        auto deflate_buffer = Compress::DeflateCompressor::compress_all(file_buffer);
-        if (!deflate_buffer.is_error() && deflate_buffer.value().size() < file_buffer.size()) {
-            member.compressed_data = deflate_buffer.value().bytes();
-            member.compression_method = Archive::ZipCompressionMethod::Deflate;
-            auto compression_ratio = (double)deflate_buffer.value().size() / file_buffer.size();
+
+        auto compression_ratio = TRY(zip_stream.add_member_from_stream(canonicalized_path, *file, date));
+        if (compression_ratio > 0.0) {
             outln("  adding: {} (deflated {}%)", canonicalized_path, (int)(compression_ratio * 100));
         } else {
-            member.compressed_data = file_buffer.bytes();
-            member.compression_method = Archive::ZipCompressionMethod::Store;
             outln("  adding: {} (stored 0%)", canonicalized_path);
```

Comment:
This `0%` is weird - if you compress a file by 90%, you would see `(deflated 10%)`. So `0%` would mean no bytes were stored at all? I'd just leave out this percentage and just show `(stored)`.